lib/boombox/bin.ex

```diff
@@ -0,0 +1,591 @@
+                  actions_acc: [],
+                  spec_builder: [],
+                  track_builders: nil,
+                  last_result: nil,
+                  eos_info: nil,
+                  rtsp_state: nil,
+                  pending_new_tracks: %{input: [], output: []},
+                  output_webrtc_state: nil
+                ]
+
+    @typedoc """
+    Statuses of the Boombox.Bin in the order of occurence.
+
+    Statuses starting with `awaiting_` occur only if
+    the `proceed` function returns `t:Wait.t/0` when in the
+    preceeding status.
+    """
+    @type status ::
+            :init
+            | :awaiting_output
+            | :output_ready
+            | :awaiting_input
+            | :input_ready
+            | :awaiting_output_link
+            | :output_linked
+            | :running
+
+    @type t :: %__MODULE__{
+            status: status(),
+            input: Boombox.Bin.input(),
+            output: Boombox.Bin.output(),
+            actions_acc: [Membrane.Bin.Action.t()],
+            spec_builder: Membrane.ChildrenSpec.t(),
+            track_builders: Boombox.Bin.track_builders() | nil,
+            last_result: Ready.t() | Wait.t() | nil,
+            eos_info: term(),
+            rtsp_state: Boombox.RTSP.state() | nil,
+            output_webrtc_state: Boombox.WebRTC.output_webrtc_state() | nil
+          }
+  end
+
+  def_input_pad :audio_input,
+    accepted_format: format when Transcoder.Audio.is_audio_format(format),
+    availability: :on_request,
+    max_instances: 1
+
+  def_input_pad :video_input,
+    accepted_format: format when Transcoder.Video.is_video_format(format),
+    availability: :on_request,
+    max_instances: 1
+
+  def_options input: [
+                spec: input() | :membrane_pad,
+                default: :membrane_pad
+              ],
+              output: [
+                spec: output()
+              ]
+
+  @impl true
+  def handle_init(ctx, opts) do
+    state = %State{
+      input: parse_endpoint_opt!(:input, opts.input),
+      output: parse_endpoint_opt!(:output, opts.output),
+      status: :init
+    }
+
+    :ok = maybe_log_transcoding_related_warning(state.input, state.output)
+
+    with :membrane_pad <- state.input, {:stream, _pid, _opts} <- state.output do
+      raise "Boombox bin cannot have pad input and stream output at the same time"
+    end
+
+    proceed(ctx, state)
+  end
+
+  @impl true
+  def handle_playing(ctx, state) when state.input == :membrane_pad do
+    cond do
+      state.status in [:init, :awaiting_output] ->
+        {[], state}
+
+      state.status == :awaiting_input ->
+        Boombox.Pad.create_input(ctx)
+        |> proceed_result(ctx, state)
+
+      true ->
+        raise "internal boombox error (state.status is #{inspect(state.status)})"
+    end
+  end
+
+  @impl true
+  def handle_playing(_ctx, state), do: {[], state}
+
+  @impl true
+  def handle_pad_added(pad_ref, ctx, state) when state.input == :membrane_pad do
```

Comment:
let's throw a readable error when someone specifies a different input and links some pads